### PR TITLE
[fix](block) fix be core while mutable block merge may cause different row size between columns in origin block

### PR DESCRIPTION
**Pull request by caiconghui**
## Proposed changes

if columns in block share column ptr, the mutable block on it will create new column ptr after merging other non-empty block, which would make origin block invalid, here is to fix it.
has different row size  

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by caiconghui** [comment]
run buildall

File: be/src/vec/exec/scan/scanner_scheduler.cpp [hunk 1]
```diff
@@ -384,10 +384,15 @@ void ScannerScheduler::_scanner_scan(ScannerScheduler* scheduler, ScannerContext
...
         } else {
             if (!blocks.empty() && blocks.back()->rows() + block->rows() <= state->batch_size()) {
-                status = vectorized::MutableBlock(blocks.back().get()).merge(*block);
+                auto rows = blocks.back()->rows() + block->rows();
+                vectorized::MutableBlock mutable_block(blocks.back().get());
+                status = mutable_block.merge(*block);
                 if (!status.ok()) {
                     break;
                 }
+                if (!blocks.back().get()->is_valid(rows)) {
```
Comment (zclllyybb) [inline be/src/vec/exec/scan/scanner_scheduler.cpp:393]: Is there any situation that it's valid that block has columns with different size? If not, maybe we can do this check in merging block.

Comment (caiconghui) [reply be/src/vec/exec/scan/scanner_scheduler.cpp:393]: the key point is that we reassign value to block from mutable block, and where to do check i think are both ok

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
#15921

**Comment by caiconghui** [comment]
run buildall

File: be/src/vec/exec/scan/scanner_scheduler.cpp [hunk 2]
```diff
@@ -394,7 +394,11 @@ void ScannerScheduler::_scanner_scan(ScannerScheduler* scheduler, ScannerContext
             ctx->return_free_block(std::move(block));
         } else {
             if (!blocks.empty() && blocks.back()->rows() + block->rows() <= state->batch_size()) {
-                static_cast<void>(vectorized::MutableBlock(blocks.back().get()).merge(*block));
+                vectorized::MutableBlock mutable_block(blocks.back().get());
+                static_cast<void>(mutable_block.merge(*block));
+                if (!blocks.back().get()->is_valid(mutable_block.rows())) {
```
Comment (HappenLee) [inline be/src/vec/exec/scan/scanner_scheduler.cpp:399]: why here need to do check ？

File: be/src/vec/core/block.cpp [hunk 3]
```diff
@@ -577,7 +586,7 @@ MutableColumns Block::mutate_columns() {
     size_t num_columns = data.size();
     MutableColumns columns(num_columns);
     for (size_t i = 0; i < num_columns; ++i) {
-        columns[i] = data[i].column ? (*std::move(data[i].column)).assume_mutable()
+        columns[i] = data[i].column ? (*std::move(data[i].column)).mutate()
```
Comment (HappenLee) [inline be/src/vec/core/block.cpp:589]: here maybe cause mutable column is different column in block，can not change the mem in block cause result errror

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run P0 Regression

**Comment by caiconghui** [comment]
run p0

**Comment by caiconghui** [comment]
run pipelinex_p0

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Comment by caiconghui** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]

**Pull request merged**